GREYErrorFormatter.m

```diff
@@ -0,0 +1,195 @@
+                         elementMatcher]];
+    }
+  }
+  
+  if ([keys containsObject:kErrorFormatterSearchActionInfoKey]) {
+    NSString *searchActionInfo = _error.userInfo[kErrorDetailSearchActionInfoKey];
+    if (searchActionInfo) {
+      [logger addObject:[NSString stringWithFormat:@"%@\n%@\n", kErrorDetailSearchActionInfoKey,
+                         searchActionInfo]];
+    }
+  }
+  
+  if ([keys containsObject:kErrorFormatterScreenshotPathsKey]) {
+    for (NSString *key in _error.appScreenshots.allKeys) {
+      NSString *screenshotPath = _error.appScreenshots[key];
+      [logger addObject:[NSString stringWithFormat:@"%@: %@\n", key, screenshotPath]];
+    }
+  }
+  
+  if ([keys containsObject:kErrorFormatterUnderlyingErrorKey]) {
+    NSString *nestedError = _error.nestedError.description;
+    if (nestedError) {
+      [logger addObject:[NSString stringWithFormat:@"%@: \n%@\n", kErrorFormatterUnderlyingErrorKey,
+                         nestedError]];
+    }
+  }
+  
+  if ([keys containsObject:kErrorFormatterUIHierarchyKey]) {
+    NSString *UIHierarchy = GREYFormattedHierarchy(_error.appUIHierarchy);
+    if (UIHierarchy) {
+      [logger addObject:UIHierarchy];
+    }
+  }
+  
+  if ([keys containsObject:kErrorFormatterStackTraceKey]) {
+    NSArray<NSString *> *stackTrace = _error.stackTrace;
+    if (stackTrace) {
+      [logger addObject:[NSString stringWithFormat:@"%@: %@\n", kErrorStackTraceKey, stackTrace]];
+    }
+  }
+  
+  return [logger componentsJoinedByString:@"\n"];
+}
+
+#pragma mark - Static Functions
+
+NSString *GREYFormattedHierarchy(NSString * hierarchy) {
+  if (!hierarchy) {
+    return nil;
+  }
+  NSMutableArray<NSString*> *logger = [[NSMutableArray alloc] init];
+  [logger addObject:kHierarchyHeaderKey];
+  NSString *windowLegend = kHierarchyWindowLegendKey;
+  NSString *axLegend = kHierarchyAcessibilityLegendKey;
+  NSString *uieLegend = kHierarchyUserInteractionEnabledLegendKey;
+  NSDictionary<NSString *, NSString *> *legendLabels = @{
+    windowLegend : kHierarchyBackWindowKey,
+    axLegend : kHierarchyAccessibilityKey,
+    uieLegend : kHierarchyUserInteractionEnabledKey
+  };
+  NSArray<NSString *> *keyOrder = @[ windowLegend, axLegend, uieLegend ];
+  NSString *legendDescription = [GREYObjectFormatter formatDictionary:legendLabels
+                                                               indent:kGREYObjectFormatIndent
+                                                            hideEmpty:NO
+                                                             keyOrder:keyOrder];
+  [logger addObject:[NSString stringWithFormat:@"%@: %@\n", kHierarchyLegendKey,
+                     legendDescription]];
+  [logger addObject:hierarchy];
+  return [logger componentsJoinedByString:@"\n"];
+}
+
+BOOL GREYShouldUseErrorFormatterForError(GREYError *error) {
+  return [error.domain isEqualToString:kGREYInteractionErrorDomain] &&
```

Comment:
This might be a nitpicking but checking error.code first then error.domain is more efficient.

Reply:
But it is always more likely to fail on the domain, since it's a string that's different for most errors, whereas the code is usually 0